### DISCUSSION TITLE
Sub-routes inside services

### DISCUSSION THREAD

**Discussion by bamidev** [tokio-rs/axum#2685] [Q&A]
### Summary

Hey,

I'm pretty new to using Axum, and I'm trying to rewrite my application to use it. I was using Rocket, but it doesn't feel right to me to have to write the same logic to load the DB data in each and every handler that uses the same URL prefix.

I had the idea of constructing the routes in such a way that a model/record gets loaded and is available in all sub-routes of the form: /actor/:hash/*
Where * will be all the subroutes which are yet to be determined.
So that I don't have to load the same model again and again in all the handlers that start with /actor/:hash

So what I did is this:
```
let app = Router::new()
  .with_state(state.clone())
  .nest_service("/static", ServeDir::new("static"))
  .nest("/actor", actor::router(state));

let listener = tokio::net::TcpListener::bind(addr).await.unwrap();
axum::serve(listener, app).await.unwrap();
```

And then in the `actor` module:
```
#[derive(Clone)]
struct ActorState {
    app: AppState,
    actor: Option<identity::Model>
}

pub fn router(app: AppState) -> Router {
    Router::new().route_service("/:hash", service_fn(move |mut req: Request| {
        let a = app.clone();
        async move {
            let hash = req.extract_parts::<Path<String>>().await.unwrap().0;

            let actor = Identity::find()
                .filter(identity::Column::Address.contains(hash))
                .one(&a.global.api.orm)
                .await.expect("unable to find actor");
            req.extensions_mut().insert(actor);
            let state = ActorState {
                app: a,
                actor: None
            };

            Router::new()
                .with_state(state)
                .route("/:hash", get(index))
                .call(req).await
        }
    }))
}

async fn index(State(state): State<ActorState>) -> impl IntoResponse {
	"HOI".to_string();
}
```

So, this doesn't compile, and it gives me this error:
```
error[E0599]: no method named `call` found for struct `axum::Router` in the current scope
  --> src/web/actor.rs:38:18
   |
35 | /             Router::new()
36 | |                 .with_state(state)
37 | |                 .route("/:hash", get(index))
38 | |                 .call(req).await
   | |                 -^^^^ method not found in `Router<ActorState>`
   | |_________________|
   | 
```

The error message is a bit hard to digest, but apparently the `tower::Service` trait isn't automatically implemented anymore. But if I remove the state argument from `index`, it does compile.
Now when I try to use the extension, it does compile, and it does work, I get the correct actor record into the index handler when it gets invoked.

So I guess the first question is: Why can't the index function for this (sub-)router take state, and then still be a service?
And second question: Is there maybe just a better way to do what I'm trying to accomplish here?

Thanks a lot in advance!

### axum version

0.7.5

**Comment by jplatte** [comment]
I don't fully understand what you're doing, but I can tell you why you're getting that error: `.with_state` supplies the state *for all routes registered so far*. You have to call it after `.route()`.

**Comment by bamidev** [comment]
Awesome, that helps me, even though I've went and tried to attack this problem from a differen angle.
I also discovered that route_layer actually only applies to the routes that are already added.
So I use that now.

I rewrote my code like this:
```
pub fn router(app: AppState) -> Router<AppState> {
    Router::new()
        .route("/:hash", get(actor_get))
        .route_layer(from_fn_with_state(app, actor_middleware))
        .route("/", get(actor_index))
}

async fn actor_middleware(
    State(state): State<AppState>,
    mut request: Request,
    next: Next,
) -> Response {
    let hash = request.extract_parts::<Path<String>>().await.unwrap().0;
    let actor = Identity::find()
        .filter(identity::Column::Address.contains(hash))
        .one(&state.global.api.orm)
        .await.expect("unable to find actor");
    request.extensions_mut().insert(actor);

    next.run(request).await
}

async fn actor_index(State(state): State<AppState>) -> impl IntoResponse {
    println!("Hello index");
	"Hello".to_string();
}

async fn actor_get(State(state): State<AppState>, Extension(actor): Extension<Option<identity::Model>>) -> impl IntoResponse {
    println!("Hello get {:?}", actor);
	"Hello".to_string();
}
```
Which compiles, and it does exactly do what I want.

The hint was appreciated!
